x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/RestAuthenticateAction.java

```diff
@@ -65,7 +70,16 @@ public RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient c
                 new RestBuilderListener<AuthenticateResponse>(channel) {
             @Override
             public RestResponse buildResponse(AuthenticateResponse authenticateResponse, XContentBuilder builder) throws Exception {
-                authenticateResponse.authentication().toXContent(builder, ToXContent.EMPTY_PARAMS);
+                if (anonymousEnabled
+                    && false == anonymousUser.equals(authenticateResponse.authentication().getUser())
+                    && anonymousUser.roles().length != 0) {
+                    builder.startObject();
+                    authenticateResponse.authentication().toXContentFragment(builder);
+                    builder.array(User.Fields.ANONYMOUS_ROLES.getPreferredName(), anonymousUser.roles());
```

Comment:
You have a point here plus it makes the whole thing easier. I have no objection. 

Just some thought process: I am ok with adding a new field. But the name "anonymous_roles" feels too specific for me to like. This does raise the question why we don't have things like `native_roles` etc. I wanted name it `inherited_roles` so it can at least to re-used. But if we treat anonymous roles just like another role providers, it should then just be part of the existing `roles` field. So yes, I'll remove the new field.